### PR TITLE
Adapter works with node >= 18

### PR DESCRIPTION
**Pull request by Arcticon**
Fixes UncleSamSwiss/ioBroker.loxone#536

### CODE REVIEW

**Comment by Arcticon** [comment]
@raintonr can i please get a quick review? :)
thanks!

**Comment by Arcticon** [comment]
@raintonr i assume we have to merge this PR first, because the node version incompatibility in the github actions is blocking other PRs.

**Comment by Arcticon** [comment]
i tried to run the tests locally with 14, 16, 18 and 20 and 18 was the only one that even started.
some iobroker dependency tries to require/import letsencrypt, but fails to - therefore throwing an error causing the startup to fail.
i can give you the log later, because im on the go right now.

**Comment by raintonr** [comment]
letsencrypt is depreciated so if there's a problem with js-controller, etc it should be fixed in that project. Did you search issues over there for the error you are seeing? If it's not known you should create a new issue.

**Comment by Arcticon** [comment]
I'll check later and will let you know :)

**Comment by Arcticon** [comment]
I have checked everything again and found out that there was an issue on my side.
Node 14 does not work, but everything from 16 on upwards works.

Node 14 log
```
$ clear; nvm exec 14 npm ci; nvm exec 14 npm run test
Running node v14.21.3 (npm v6.14.18)
npm WARN prepare removing existing node_modules/ before installation
npm ERR! Cannot read property '@iobroker/adapter-core' of undefined

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/2024-05-13T07_23_36_405Z-debug.log
Running node v14.21.3 (npm v6.14.18)

> iobroker.loxone@3.0.1 test dev/ioBroker.loxone
> npm run test:ts && npm run test:package


> iobroker.loxone@3.0.1 test:ts dev/ioBroker.loxone
> mocha --config test/mocharc.custom.json src/**/*.test.ts

sh: 1: mocha: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! iobroker.loxone@3.0.1 test:ts: `mocha --config test/mocharc.custom.json src/**/*.test.ts`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the iobroker.loxone@3.0.1 test:ts script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/2024-05-13T07_23_37_300Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! iobroker.loxone@3.0.1 test: `npm run test:ts && npm run test:package`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the iobroker.loxone@3.0.1 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/2024-05-13T07_23_37_309Z-debug.log
```

**Review by raintonr — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export type StateChangeListenEntry = {
     listener: StateChangeListener;
     opts?: StateChangeListenerOpts;
     queuedVal: ioBroker.StateValue | null;
-    ackTimer: ioBroker.Timeout | null;
+    ackTimer: ioBroker.Timeout | undefined;
```
[the remark this answers is not in the document]
Comment (Arcticon) [reply src/main.ts:36]: i agree, but the update to `adapter-core@3.1.4` changes the function declaration to:

```
AdapterClass.setTimeout(cb: TimeoutCallback, timeout: number, ...args: any[]): ioBroker.Timeout | undefined
```

prior to this change it returned `null`

File: src/main.ts [hunk 2]
```diff
@@ -384,10 +384,12 @@ export class Loxone extends utils.Adapter {
             if (!stateChangeListener.opts?.selfAck) {
                 // Set ack timer before calling listener
                 stateChangeListener.ackTimer = this.setTimeout(
+                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                    // @ts-ignore
```
Comment (Arcticon) [inline src/main.ts:388]: i had to add a `@ts-ignore` here, because the type of `setTimeout` is not correct

Comment (Arcticon) [reply src/main.ts:388]: they defined the TimeoutCallback like so:
`export declare type TimeoutCallback = (args?: any[]) => void;`
but it should be like this:
`export declare type TimeoutCallback = (...args: any[]) => void;`

File: src/main.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (raintonr) [reply src/main.ts:36]: Hmmmmm.... https://github.com/ioBroker/ioBroker.js-controller/pull/2375/files#diff-fdf15c6340b088e24a291222e4d17edc0b96c2dde90b77d0f2a499dd24fcd425L2590

Fair enough then.

File: src/main.ts (hunk 2 above)
Comment (raintonr) [reply src/main.ts:388]: If you think this is incorrect please raise it as an issue on that project.

**Review by raintonr — APPROVED** [review]

File: src/main.ts (hunk 2 above)
Comment (Arcticon) [reply src/main.ts:388]: will do

**Comment by Arcticon** [comment]
@raintonr the tests are failing because the latest js controller requires node 18:
https://github.com/ioBroker/ioBroker.js-controller/blob/v5.0.20-alpha.0-20240510-819f1976e/packages/controller/package.json#L6

File: src/main.ts (hunk 2 above)
Comment (Arcticon) [reply src/main.ts:388]: Link: https://github.com/ioBroker/ioBroker.js-controller/issues/2733

**Comment by Arcticon** [comment]
Reported that as well: https://github.com/ioBroker/ioBroker.js-controller/issues/2734

**Comment by Arcticon** [comment]
js-controller 5.0.19 works with node 16. will manifest that

**Comment by Arcticon** [comment]
@raintonr could you please trigger the pipeline again?

**Comment by Arcticon** [comment]
@raintonr do you want to wait for the fix for the `ts-ignore`'s or should this be "ok" and i can create a follow up issue?

**Comment by mcm1957** [comment]
Please adapt test-and-release to current woirkflow using @iobroker/testing.
Standard tests are required to pass.
Node 16 must be removed as it is no longer supported by standard tests.
engines should be raise to node >=18

Note that I will not release any new release to stable unless it passes standard github tests.

**Comment by raintonr** [comment]
> @raintonr do you want to wait for the fix for the `ts-ignore`'s or should this be "ok" and i can create a follow up issue?

Nah, let's not wait. Although please remove node 16 as per @mcm1957 comment above. Then think it's good to go. Thanks for your help.

**Comment by Arcticon** [comment]
@raintonr have done :)

**Pull request merged**